pkg/controller/deployment_rolling.go

```diff
@@ -341,3 +367,112 @@ func (dc *controller) taintNodesBackingMachineSets(MachineSets []*v1alpha1.Machi
+
+	for _, machineSet := range MachineSets {
+
+		klog.V(3).Infof("Trying to annotate nodes under the MachineSet object %q with %s", machineSet.Name, annotations)
+		selector, err := metav1.LabelSelectorAsSelector(machineSet.Spec.Selector)
+		if err != nil {
+			return err
+		}
+
+		// list all machines to include the machines that don't match the ms`s selector
+		// anymore but has the stale controller ref.
+		// TODO: Do the List and Filter in a single pass, or use an index.
+		filteredMachines, err := dc.machineLister.List(labels.Everything())
+		if err != nil {
+			return err
+		}
+		// NOTE: filteredMachines are pointing to objects from cache - if you need to
+		// modify them, you need to copy it first.
+		filteredMachines, err = dc.claimMachines(machineSet, selector, filteredMachines)
+		if err != nil {
+			return err
+		}
+
+		for _, machine := range filteredMachines {
+			if machine.Status.Node != "" {
+				err = AddOrUpdateAnnotationOnNode(
+					dc.targetCoreClient,
+					machine.Status.Node,
+					annotations,
+				)
+				if err != nil {
+					klog.Warningf("Adding annotation failed for node: %s, %s", machine.Status.Node, err)
+				}
+			}
+		}
+		klog.V(2).Infof("Annotated the nodes backed by MachineSet %q with %s", machineSet.Name, annotations)
+	}
+
+	return nil
+}
+
+func (dc *controller) machineSetsScaledToZero(MachineSets []*v1alpha1.MachineSet) bool {
+	for _, machineSet := range MachineSets {
+		if machineSet.Spec.Replicas != 0 && machineSet.Status.AvailableReplicas != 0 && machineSet.Status.FullyLabeledReplicas != 0 && machineSet.Status.ReadyReplicas != 0 && machineSet.Status.Replicas != 0 {
+			return false
+		}
+	}
+	return true
+}
+
+// removeAutoscalerAnnotationsIfRequired removes the annotations if needed from nodes backing machinesets.
+func (dc *controller) removeAutoscalerAnnotationsIfRequired(MachineSets []*v1alpha1.MachineSet, annotations map[string]string) error {
+
+	for _, machineSet := range MachineSets {
+
+		selector, err := metav1.LabelSelectorAsSelector(machineSet.Spec.Selector)
+		if err != nil {
+			return err
+		}
+
+		// list all machines to include the machines that don't match the ms`s selector
+		// anymore but has the stale controller ref.
+		// TODO: Do the List and Filter in a single pass, or use an index.
+		filteredMachines, err := dc.machineLister.List(labels.Everything())
+		if err != nil {
+			return err
+		}
+		// NOTE: filteredMachines are pointing to objects from cache - if you need to
+		// modify them, you need to copy it first.
+		filteredMachines, err = dc.claimMachines(machineSet, selector, filteredMachines)
+		if err != nil {
+			return err
+		}
+
+		for _, machine := range filteredMachines {
+			if machine.Status.Node != "" {
+
+				nodeAnnotations, err := GetAnnotationsFromNode(
+					dc.targetCoreClient,
+					machine.Status.Node,
+				)
+				if err != nil {
```

Comment:
should we ignore "not found" error?

Reply:
thanks, it's a good point. I have basically taken the methods from a similar mechanism of taints already implemented- check [here](https://github.com/gardener/machine-controller-manager/blob/master/pkg/controller/controller_utils.go#L900-L937).

But I see a value in ignoring the node-not-found error. I believe, otherwise, the controller will constantly retry to find the node, a bit of load on apiserver. 
@prashanth26  Did you see any reason for not handling the `NotFound` errors for taints as well back then. Else, I am happy to patch other methods related to taints[+conditions] here.

Reply:
Done for the annotation [here](https://github.com/gardener/machine-controller-manager/pull/496/commits/1c1543420a6b3b89c78079aabedcb49d6bd862e4) . 

- A side thought would then be, that the machine will be ignored for further reconciliation if the node-object is going to join in sometime. I hope the event for node's joining should trigger the reconciliation later.

Reply:
This method was already taking care of it, right? - https://github.com/gardener/machine-controller-manager/pull/496/files#diff-44bba2c68a6ea2b4ca0dbb7f4465828eR1231-R1234. No?

Reply:
I added it just now :D 
There are some methods for taints implemented earlier, which doesn't handle the not-found case - https://github.com/gardener/machine-controller-manager/blob/master/pkg/controller/controller_utils.go#L900-L937 .
- I meant if we should update these methods already or keep it for later.